v1beta OpenAPI-2.0 specification & go-swagger code generation

tgraf opened the pull request:
Major rework of the API including specification of the API with swagger-2.0, code generation based on it and all required conversion to use the generated code.

aanm commented on daemon/daemon/endpoint.go:
```diff
@@ -26,6 +26,8 @@ import (
 	"github.com/cilium/cilium/pkg/labels"
 	"github.com/cilium/cilium/pkg/option"
 	"github.com/cilium/cilium/pkg/policy"
+
```
Remove line

aanm commented on api/v1/client/cilium_client.go:
```diff
@@ -0,0 +1,78 @@
+package client
```
Out of curiosity, is not possible to have headers?

tgraf replied:
Not sure what you mean here, can you explain?

aanm replied:
Sorry, the license headers on the created packages.

tgraf replied:
I think this is a non trivial matter. I want to understand this better before we blindly attach a license to generated code.

aanm commented on cli/endpoint.go:
```diff
@@ -232,66 +210,47 @@ func printEndpointLabels(lbls *labels.OpLabels) {
-	if err := client.EndpointLabelsUpdate(epID, lo); err != nil {
-		log.Errorf("Error while deleting labels %s", err)
+	if err := client.EndpointLabelsPut(id, "cilium", lo); err != nil {
+		log.Errorf("Error while modiyfing labels %s", err)
```
modifying

tgraf replied:
Done

aanm commented on cli/endpoint.go:
```diff
@@ -232,66 +210,47 @@ func printEndpointLabels(lbls *labels.OpLabels) {
 	}
 
-	if lbls, err := client.EndpointLabelsGet(epID); err != nil {
+	if lbls, err := client.EndpointLabelsGet(id, "cilium"); err != nil {
```
What does "cilium" means?

tgraf replied:
It corresponds to the cilium endpoint ID. I will move this to a constant.

aanm commented on cli/policy.go:
```diff
@@ -365,9 +339,12 @@ func importPolicy(ctx *cli.Context) {
+		if resp, err := client.PolicyPut(node.Name, jsonPolicy); err != nil {
 			fmt.Fprintf(os.Stderr, "Could not import policy directory %s: %s\n", path, err)
 			os.Exit(1)
+		} else if ctx.Bool("echo") {
```
Before I forget, please insert a OK message here if the policy was added.
```
		} else {
			fmt.Printf("Policy successfully inserted\n")
			if ctx.Bool("echo") {
				fmt.Printf("%s\n", resp)
			}
  		}
```

tgraf replied:
Not sure that's a good idea for use in scripts. I've added the `-e` flag to trigger output if successful and we have the exit code to check for success/failure.

Do you see a particular reason for a success message?

aanm replied:
When we are importing a policy with the CLI we don't have any feedback what happen with the operation. What about `fmt.Fprintf(os.Stderr, "Policy...")`?

tgraf replied:
The feedback is a zero or non-zero exit status plus verbose error information on stderr in case of the latter. If a user wants to have output on stdout on success he can use `--echo`

aanm commented on common/types/loadbalancer.go:
```diff
@@ -57,17 +81,30 @@ type RevNATMap map[ServiceID]L3n4Addr
 
+func NewL4Type(name string) (L4Type, error) {
+	switch name {
+	case "tcp":
```
`case "tcp", "TCP":` maybe?

tgraf replied:
Good point. I've added a `strings.ToLower()`

aanm commented on daemon/daemon/daemon.go:
```diff
@@ -523,22 +526,73 @@ func (d *Daemon) staleMapWalker(path string) error {
 
 	if err := d.conf.Opts.Validate(opts); err != nil {
-		return err
+		return NewApiError(400, err)
```
Golang has http errors, `http.StatusBadRequest`

tgraf replied:
I'm not quite happy with the error handling yet. Ideally all errors are defined in the API spec so users can act upon particular errors with well known error codes. Doing so well generate a type for each and the client side will convert it back to a type with a corresponding name, e.g. `PutEndpointAlreadyExists`

aanm commented on daemon/daemon/daemon.go:
```diff
@@ -523,22 +526,73 @@ func (d *Daemon) staleMapWalker(path string) error {
-			log.Warningf("Unable to recompile base programs: %s\n", err)
+			msg := fmt.Errorf("Unable to recompile base programs: %s\n", err)
+			log.Warningf("%s", msg)
+			return NewApiError(401, msg)
```
Golang has http errors, `http.StatusUnauthorized`

aanm commented on daemon/daemon/endpoint.go:
```diff
@@ -203,35 +209,67 @@ func (d *Daemon) createBPFMAPs(epID uint16) error {
-func (d *Daemon) EndpointLeave(epID uint16) error {
 	d.endpointsMU.Lock()
-	defer d.endpointsMU.Unlock()
+	existingEp, ok := d.endpoints[ep.ID]
```
Isn't this a pointer? Doesn't this creates concurrency problems for using it outside the d.endpointsMU.Unlock?

tgraf replied:
We only read from it but you are right that we have a concurrency issue in general here. We realy on create/join which are decoupled and multiple writers can modify them concurrently. This possibility existed before. I started introducing states to ensure that transitions can only happen in a well defined order but we need to improve this.

aanm commented on daemon/daemon/endpoint.go:
```diff
@@ -260,124 +298,218 @@ func (d *Daemon) EndpointLeave(epID uint16) error {
 
-func (d *Daemon) EndpointLabelsUpdate(epID uint16, labelOps labels.LabelOp) error {
-	ep, err := d.EndpointGet(epID)
+func (d *Daemon) filterLabels(from []string) labels.Labels {
```
Add a comment here `Must be called with d.conf.ValidLabelPrefixMU RLocked`

tgraf replied:
Done

aanm commented on daemon/daemon/loadbalancer.go:
```diff
@@ -140,16 +218,24 @@ func (d *Daemon) SVCDelete(feL3n4 types.L3n4Addr) error {
-		delete(d.loadBalancer.SVCMap, feL3n4Uniq)
+
+	if err := lbmap.DeleteService(svcKey); err != nil {
+		return nil
```
You mean `return err`, right?

aanm replied:
Fixed

aanm commented on daemon/main.go:
```diff
@@ -19,20 +19,25 @@ import (
 	"github.com/cilium/cilium/pkg/endpoint"
 	"github.com/cilium/cilium/pkg/labels"
 	"github.com/cilium/cilium/pkg/option"
+	loads "github.com/go-openapi/loads"
```
new line here

tgraf replied:
Seems to have been caused by latest merge. Fixed.

aanm commented on pkg/client/client.go:
```diff
@@ -22,44 +22,51 @@ import (
 
+	clientapi "github.com/cilium/cilium/api/v1/client"
 	common "github.com/cilium/cilium/common"
-
```
new line

tgraf replied:
Seems to have been caused by latest merge. Fixed.

aanm commented on pkg/endpoint/policy.go:
```diff
@@ -43,9 +43,9 @@ func (e *Endpoint) checkEgressAccess(owner Owner, opts option.OptionMap, dstID p
 	switch owner.GetPolicyTree().Allows(&ctx) {
 	case policy.ACCEPT, policy.ALWAYS_ACCEPT:
-		opts[opt] = true
+		opts[opt] = "enabled"
```
Why strings instead of booleans?

tgraf replied:
So we can carry more than just boolean's. Things like the size of the connection tracking table, and so on. Undecided yet whether we should hardcode those things in the API itself as done for some other settings now or treat it as a generic key value map.

aanm commented on plugins/cilium-cni/cilium-cni.go:
```diff
@@ -238,75 +284,83 @@ func cmdAdd(args *skel.CmdArgs) error {
+
+	res := cniTypes.Result{}
+
+	if ipam.IPV6 != "" {
```
We need an endpoint ID from ipam.IPV6, if it's empty, should we fail and return errors?

tgraf replied:
Probably the right this to do for this PR. I'm looking at getting rid of this requirement but I'll defer that to a separate PR.

tgraf replied:
Fixed

aanm commented on plugins/cilium-cni/cilium-cni.go:
```diff
@@ -238,75 +284,83 @@ func cmdAdd(args *skel.CmdArgs) error {
 	}
 
-	ep, err := c.EndpointGetByDockerID(args.ContainerID)
+	ep, err := client.EndpointGet(args.ContainerID, "docker-id")
```
Can't we replace those strings ("docker-id") with an enum?

tgraf replied:
Yes, will do

borkmann commented on api/v1/openapi.json:
```diff
@@ -0,0 +1,1115 @@
+                                        "200": {
+                                                "description": "Success",
+                                                "schema": {
+                                                        "$ref": "#/definitions/StatusResponse"
```
Could this get extended in future to include stats (f.e. from maps), etc, or should this be separated under a different path?

tgraf replied:
The returned json object can be extended in the future to include additional stuff.

borkmann commented on api/v1/openapi.json:
```diff
@@ -0,0 +1,1115 @@
+                                "summary": "Get health of Cilium daemon",
+                                "description": "Represents the runtime status of the daemon and related components.",
+				"tags": [ "daemon" ],
+                                "responses": {
```
Nit: seems indented differently (also elsewhere).

tgraf replied:
Fixed

borkmann commented on api/v1/openapi.json:
```diff
@@ -0,0 +1,1115 @@
+        "info": {
+                "title": "Cilium API",
+                "description": "Cilium",
+                "version": "unversioned"
```
Should we make this "0.1" or something? Does it have any side-effects to leave it as "unversioned" or would we rather update the base path later on (when necessary)?

borkmann replied:
Do we need "license" member for the exposed API?

tgraf replied:
Will change this to v1beta to make it clear that this is being stabilized towards v1. I'm not sure an API can even be licensed in this regard so I've omitted that for now.

borkmann commented on api/v1/openapi.json:
```diff
@@ -0,0 +1,1115 @@
+                "http"
+	],
+        "x-schemes": [
+		"unix"
```
The x-schemes vendor extension is needed for json, right?

tgraf replied:
x-schemes is a go-swagger extension to allow unix domain sockets in the generated code.

borkmann commented on api/v1/openapi.json:
```diff
@@ -0,0 +1,1115 @@
+                                }
+                        }
+                },
+                "/config": {
```
Should we rather make this /daemon/config?

tgraf replied:
Fair question. The question is whether that would apply to `/ipam/` and `/policy/` as well then.

borkmann replied:
I guess it would depend on whether we think in future there could be similar paths for other cilium contexts where we would potentially run into collision. I'm fine either way, was just a thought.

borkmann commented on api/v1/openapi.json:
```diff
@@ -0,0 +1,1115 @@
+                        "description": "List of configuration settings",
+			"type": "object",
+			"additionalProperties": {
+				"type": "string"
```
Should this be more specific here or not possible since the data is contained in body?

tgraf replied:
This is a generic key value representation

borkmann commented on api/v1/openapi.json:
```diff
@@ -0,0 +1,1115 @@
+		"Endpoint": {
+			"description": "Endpoint",
+                        "type": "object",
+			"properties": {
```
Should an endpoint also contain the compiled-in features from bpf side (f.e. nat64 support, etc)?

tgraf replied:
This is represented by the `/endpoint/config/`. The question is whether this in fact needs to be a separate API path. It's separate right now to distinguish between create/delete/state-change of endpoints and mere configuration changes.

borkmann commented on api/v1/openapi.json:
```diff
@@ -0,0 +1,1115 @@
+				"docker-network-id": {
+					"description": "Docker network ID",
+					"type": "string"
+				},
```
Should we wrap all ids into a single "id" object with subtypes? Also, all the ids might be confusing with "identity" later on. Maybe some restructuring would be good?

Perhaps all docker related info could go into "docker": { "container": { "id": ... }, "endpoint": { "id": ... }, "network": { "id": ... } }. Then for cilium, we could perhaps have "id" as is. And, further below, should we restructure this into "security": {"identity": {...}} and "network": {...} where we put all network properties like "interface": {"name":{}, "index":{}}, etc? I think it would then have a better structure perhaps.

tgraf replied:
Agreed. will rework this.

borkmann commented on api/v1/openapi.json:
```diff
@@ -0,0 +1,1115 @@
+				"ipv4": {
+					"description": "IPv6 address",
+					"type": "string"
+				},
```
Could "ipv4"/"ipv6" from here and below in IPAM be made as one definition and reused for each?

tgraf replied:
Fixed

borkmann commented on api/v1/openapi.json:
```diff
@@ -0,0 +1,1115 @@
+                                                        "$ref": "#/definitions/Error"
+                                                }
+                                        },
+                                        "default": {
```
Should we have an error that says that we've no more resources to create a new endpoint or does this fall under "default"?

tgraf replied:
It currently falls under default but this is a good case for an error that must be distinguishable

tgraf replied:
Fixed

borkmann commented on api/v1/openapi.json:
```diff
@@ -0,0 +1,1115 @@
+				],
+                                "responses": {
+					"200": {
+						"description": "Successful update"
```
Should we say "Success" everywhere, iow make it consistent?

tgraf replied:
Fixed

borkmann commented on api/v1/openapi.json:
```diff
@@ -0,0 +1,1115 @@
+						}
+					},
+                                        "405": {
+                                                "description": "Endpoint already exists",
```
The summary says "Create or update an endpoint". Do we have a way to force an update on an existing EP or is currently only creation supported? In latter case, maybe the description would need an update?

borkmann commented on api/v1/openapi.json:
```diff
@@ -0,0 +1,1115 @@
+                                                "description": "Endpoint not found",
+                                                "schema": {
+                                                        "$ref": "#/definitions/Error"
+                                                }
```
Is deletion atomic (wrt parallel requests)? In other words would we need some state 'deletion in progress'?

tgraf replied:
Deletion is imminent and atomic. Creation is not atomic because there is the state of creation the endopint, then retrieving of metadata followed by program generation and attachment until ready. Will document this.

borkmann commented on api/v1/openapi.json:
```diff
@@ -0,0 +1,1115 @@
+                                                }
+					}
+				}
+			}
```
Could we also batch-create a list of EPs via "put" here?

tgraf replied:
We can add support for this via `PUT /endpoint/batch/` which assumes an array of endpoints and returns an array of return codes. I think it makes sense to keep this separate from the existing operation as it will require a policy argument of sort to define whether to continue creating up error.

borkmann replied:
Makes sense.

borkmann commented on api/v1/openapi.json:
```diff
@@ -0,0 +1,1115 @@
+                                "responses": {
+					"200": {
+						"description": "Successful update"
+					},
```
Do we need "404" error here as well?

tgraf replied:
Fixed

borkmann commented on api/v1/openapi.json:
```diff
@@ -0,0 +1,1115 @@
+                                                "schema": {
+                                                        "$ref": "#/definitions/Error"
+                                                }
+                                        }
```
Do we need "default" error case (also everywhere else where we don't have it explicitly yet)?

tgraf replied:
It is specifically left out if no unexpected error may occur, e.g. `GET `/endpoint/labels/` is either success or 404. There are no other errors. Will double check this everywhere though.

borkmann commented on api/v1/openapi.json:
```diff
@@ -0,0 +1,1115 @@
+						"in": "body",
+						"required": true,
+						"schema": {
+							"$ref": "#/definitions/ConfigurationMap"
```
Should the ConfigurationMap also be included on EP creation request itself further above?

borkmann replied:
Where can we set the immutable config on EP creation?

tgraf replied:
Definitely something we should add, this is currently not possible. The API can be extended with this later on though without breaking compatibility.

borkmann commented on api/v1/openapi.json:
```diff
@@ -0,0 +1,1115 @@
+                                                "description": "Endpoint not found",
+                                                "schema": {
+                                                        "$ref": "#/definitions/Error"
+                                                }
```
Could we also have EPs with immutable labels in future? And we'd get EPERM?

tgraf replied:
Good point, will add this error once we have this restriction.

borkmann commented on api/v1/openapi.json:
```diff
@@ -0,0 +1,1115 @@
+						"name": "container-id",
+						"in": "query",
+						"type": "string"
+					}
```
Does order matter here, iow, do we need to put container-id first?

borkmann replied:
Do we need "404" when intention is to update an existing container-id with label set?

tgraf replied:
Documented the current behaviour in detail.

borkmann commented on api/v1/openapi.json:
```diff
@@ -0,0 +1,1115 @@
+		},
+                "/identity/max-id": {
+			"get": {
+				"summary": "Retrieve the maximum identity ID",
```
Do you have a use-case where this is needed? This is the max-id over all EPs? Should this just be named "max"?

tgraf replied:
We are currently exporting this via the API but I'm not sure it fulfills an actual purpose. We use it internally to be able to iterate over all identities. I'll remove it from the API.

borkmann replied:
Yeah, sounds good.

borkmann commented on api/v1/openapi.json:
```diff
@@ -0,0 +1,1115 @@
+                                                "schema": {
+                                                        "$ref": "#/definitions/IPAM"
+                                                }
+                                        },
```
Should we add an error if an addr is in use already?

tgraf replied:
Fixed

borkmann commented on api/v1/openapi.json:
```diff
@@ -0,0 +1,1115 @@
+		},
+                "/ipam": {
+			"get": {
+				"summary": "Allocate an IP address",
```
Hmm, is this summary correct for "get"? Seems more for "put" part?

tgraf replied:
It is correct although the question is whether this should in fact be a PUT operation. It allocates an IP address out of the pool and returns it.

tgraf replied:
I agree that `GET` - `DELETE` is an odd pair for allocate/release. `PUT` - `DELETE` makes more sense. Fixed.

borkmann commented on api/v1/openapi.json:
```diff
@@ -0,0 +1,1115 @@
+		},
+                "/ipam/{ip}": {
+			"get": {
+				"summary": "Allocate an IP address",
```
Ditto?

tgraf replied:
Fixed

borkmann commented on api/v1/openapi.json:
```diff
@@ -0,0 +1,1115 @@
+				"responses": {
+					"200": {
+                                                "description": "Success"
+                                        }
```
"404" error?

tgraf replied:
Fixed

borkmann commented on api/v1/openapi.json:
```diff
@@ -0,0 +1,1115 @@
+                                                "schema": {
+							"$ref": "#/definitions/PolicyTree"
+                                                }
+					},
```
Should we add more elaborate error codes here wrt misconfiguration?

tgraf replied:
Yes, I think this should include a `PolicyValidationError` which includes type of syntax error, line number, etc. Will add this.

borkmann commented on api/v1/openapi.json:
```diff
@@ -0,0 +1,1115 @@
+					{ "$ref": "#/parameters/service-config" }
+				],
+				"responses": {
+					"201": {
```
Should this be "200"?

borkmann replied:
Likewise "404" error when updating existing service?

tgraf replied:
`201` is the HTTP status for created. I have changed this consistently everywhere now.

tgraf replied:
I've changed it to `200` for update successful and `201` for created to see whether create or update was performed.

borkmann commented on api/v1/openapi.json:
```diff
@@ -0,0 +1,1115 @@
+			"enum": [
+				"cilium",
+				"docker-id",
+				"docker-endpoint"
```
See comment on "id" from below. Could we make that more consistent or better structured? F.e. "docker-id" vs. "docker-endpoint" as "endpoint-id-type" seems perhaps weird. "cilium" is mentioned here but implicitly id in the EP. If we could formalize that in a more structured way, would be great.

borkmann commented on api/v1/openapi.json:
```diff
@@ -0,0 +1,1115 @@
+				"enabled": {
+                                        "$ref": "#/definitions/Labels"
+				},
+				"probes": {
```
What is "probes"?

tgraf replied:
Andre needs to explain ;-) This needs to be clarified and properly documented.

borkmann commented on api/v1/openapi.json:
```diff
@@ -0,0 +1,1115 @@
+			"properties": {
+				"ip": {
+					"description": "Layer 3 address",
+					"type": "string"
```
Do we need to make this more fine-grained here and in "FrontendAddress" wrt differentiating ipv4/ipv6?

tgraf replied:
The string can contain either an IPv4 or IPv6 address

aanm reviewed (commented):
Can you change `api/v1/openapi.yaml` back to json? yaml is difficult to read on a huge file like this.

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,932 @@
+    properties:
+      all:
+        "$ref": "#/definitions/Labels"
+      user:
```
description: Labels inserted by users on an endpoint using cilium CLI or cilium related tools.

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,932 @@
+        "$ref": "#/definitions/Labels"
+      enabled:
+        "$ref": "#/definitions/Labels"
+      probes:
```
description: Labels retrieved from probes that are updated from external events such as docker and kubernetes.

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,932 @@
+        "$ref": "#/definitions/Labels"
+      user:
+        "$ref": "#/definitions/Labels"
+      enabled:
```
description: Labels enabled on the endpoint. An endpoint have multiple labels from multiple sources but the user might want to enable only one of those labels at a particular time.

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,932 @@
+    description: Label configuration of an endpoint
+    type: object
+    properties:
+      all:
```
description: All labels, used to distinguish which probes labels can be disabled/deleted by the user without them to reappearing next time a probe pools the same labels.

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+    delete:
+      summary: Delete endpoint
+      description: |
+        Deletes the endpoint specified by the ID. Deletion is imminent and
```
Reading this description made me think, if the container is running should we allow a deletion without a `--force` flag?

tgraf replied:
Such a flag can be done in the CLI. No need to put this in the API.

aanm replied:
But don't you need to have it in the API?

tgraf replied:
Why?

aanm replied:
How does 3rd parties know there's a "force" option available by only reading the docs?

tgraf replied:
If you use an API to delete an endpoint then you need to know what you are doing. This is different from a `--force` option on the CLI where users just run random commands. Any such force functionality should be implemented in whatever calls the API.

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+        '200':
+          description: Success
+        '206':
+          description: Deleted with a number of errors encountered
```
Only the number of errors? Not their details?

tgraf replied:
Can always be added but I want to first try and get rid of the error conditions so this is a workaround for an inconsistent API execution for now.

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+          schema:
+            type: integer
+        '400':
+          description: Invalid endpoint ID format for specified type. Details in error message"
```
I believe there's a forgotten `"` at the end of this sentence.

tgraf replied:
Fixed

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+        '404':
+          description: Endpoint not found
+  "/endpoint":
+    put:
```
Why not `post`? If you're creating an object without knowing their final URI I believe it should be a POST and let the server decide what'll be their URL. If the user would need to update an endpoint should use  `PUT: endpoint/{id}`.

tgraf replied:
Sure, fixed

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+        Updates an existing endpoint or creates it if it does not exist.
+        In order to update an endpoint, the ID field of the endpoint
+        must be set. The return code 200 and 201 indicates whether an
+        existing endpoint was updated or a new endpoint was created.
```
OK now I'm having mix feelings about my comment in line 121 :-)

tgraf replied:
We can have a POST on /endpoint and PUT on /endpoint/{ID} but does not allow to have create-or-update

tgraf replied:
Actually, update is not implemented yet and I'm not sure which part outside of labels and config should be updateable. I'll revise the description and remove the update part.

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+          schema:
+            "$ref": "#/definitions/Error"
+        '460':
+          description: No endpoint information in request
```
If the endpoint is `required: true` will the server even receive a request without an endpoint on his side? I mean, won't swagger simply drop the request if it doesn't contain an endpoint?

tgraf replied:
Just validated this, yes. Generated code enforces this. Removed.

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+          description: No endpoint information in request
+          x-go-name: Missing
+        '461':
+          description: Can only join existing endpoint in created state
```
What you mean by join? It's the first time I'm reading this term in the documentation.

tgraf replied:
I'll explain it. It means joining a network.

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+          description: Endpoint already exists
+          x-go-name: Exists
+        '500':
+          description: Join failed
```
Join here again.

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+            "$ref": "#/definitions/Error"
+        '507':
+          description: Out of resources
+          x-go-name: OutOfResources
```
Won't adding the schema help to know which resources we are out of?
```
          schema:
            "$ref": "#/definitions/Error"
```

tgraf replied:
Yes, this is not implemented yet but once we have it we can return a schema that represents this.

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+        Returns an array of all local endpoints managed by the daemon.
+      tags:
+      - endpoint
+      responses:
```
and a `204` if there are no endpoints maybe?

tgraf replied:
Advantage over an empty array?

aanm replied:
Thought about it a little bit, it's an unprobable unnecessary unmarshalling on the client side vs conditional if == 204 on every response?

tgraf replied:
Let's not optimize prematurely. This can always be added later on if really required. The smaller the API, the better.

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+        '404':
+          description: Endpoint not found
+        '460':
+          description: Label to be deleted not found
```
Add something like: `if this error occurs the previous labels' endpoint where not modified`

tgraf replied:
Fixed

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+            "$ref": "#/definitions/Identity"
+        '404':
+          description: Identity not found
+        '520':
```
502 or 503 maybe?
```
502 Bad Gateway
The server was acting as a gateway or proxy and received an invalid response from the upstream server.
503 Service Unavailable
The server is currently unavailable (because it is overloaded or down for maintenance). Generally, this is a temporary state.
```

Note: the same goes for Identity

tgraf replied:
I thought about that but they are quite different. Both are interpreted by HTTP proxies so I want to avoid them.

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+          schema:
+            "$ref": "#/definitions/Error"
+        '521':
+          description: Invalid identity format in storage
```
When/How can this happen? If this happen can we try to fix it without telling the user?

Note: the same goes for Identity

tgraf replied:
It would likely imply that Cilium was upgraded and the storage has a different schema version which is why I wanted to have a distinct error condition for this.

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+      summary: Create or update an identity
+      description: |
+        Creates a new identity for the specified list of labels which are
+        assumed to uniquely define the identity. If the labels have been
```
`If the labels have been` should be `If the labels are being used`. Then in the end of the sentence `Labels are considered in unused if all of their objects (containers IDs) weren't seen after a timeout`

tgraf replied:
fixed

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+        ensure the ID associated with the identity is unique in the entire
+        cluster.
+
+        If the optional container-id parameter is provided, the container ID
```
I think we should specify here `optional global endpoint ID` I want to mean that we are not container specific

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+        required: true
+        schema:
+          "$ref": "#/definitions/Labels"
+      - name: container-id
```
I think we should specify here `optional global endpoint ID` I want to mean that we are not container specific

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+      - name: container-id
+        description: |
+          Optional container ID to be associated with the identity
+        in: query
```
If this is in query maybe we should point out there's a particular [set of rules](https://tools.ietf.org/html/rfc3986#section-2.2)

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+          description: Success
+          schema:
+            "$ref": "#/definitions/IPAM"
+        '401':
```
Maybe '500', 4xx are for client errors and allocation failure seems a server error.

Also we might have '507' in case we get out of free IPs to give.

```
507 Insufficient Storage (WebDAV; RFC 4918)
The server is unable to store the representation needed to complete the request.[15]
```

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+          schema:
+            "$ref": "#/definitions/Error"
+  "/ipam":
+    put:
```
POST sounds better on this case.

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+          x-go-name: InvalidPolicy
+          schema:
+            "$ref": "#/definitions/Error"
+        '460':
```
What's the reason for not having 460 and 461 be used in 400? We could use the "Invalid policy" description and the error would describe what was the precise error: "Missing policy" or "Invalid path" for example.

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+          description: Policy tree not found
+        '500':
+          description: Error while deleting policy
+          x-go-name: Failure
```
Missing schema?
```
          schema:
            "$ref": "#/definitions/Error"
```

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+      tags:
+      - service
+      responses:
+        '200':
```
`204` if empty?

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+            "$ref": "#/definitions/Service"
+        '404':
+          description: Service not found
+    put:
```
It's not consistent, you previously have `put /identity` here you have `put /service/{id}`. You can still have `put /service/{id}` for updates and you should add `post /service` for creation.

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+    type: integer
+  identity-hash:
+    name: hash
+    description: SHA256 hash of identity labels
```
In base64 right?

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+      "$ref": "#/definitions/IdentityContext"
+  service-id:
+    name: id
+    description: ID of service
```
Cluster wide?

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+    type: object
+    properties:
+      id:
+        description: Unique identification
```
Cilium unique identification. Is a global unique identification?

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+        description: MAC address
+        type: string
+      host-mac:
+        description: MAC address
```
MAC address of host side interface.

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+        description: Security identity
+        "$ref": "#/definitions/Identity"
+  EndpointResponse:
+    description: Response to an endpoint PUT request
```
Why an endpoint PUT request returns an IPAM configuration?

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+      delete:
+        description: |
+          List of labels to delete. If the label was derived from an
+          orchestration system. It will not be deleted but put onto the
```
"orchestration system. It will" -> "orchestration system it will"

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+        description: Status of key/value store
+        "$ref": "#/definitions/Status"
+      docker:
+        description: Status of local container runtime
```
You're right, "local container runtime" sounds better perhaps we should rename the property from "docker" to "container-runtime"?

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+        description: Status of logstash integration
+        "$ref": "#/definitions/Status"
+      ipam:
+        description: Status of IP address management
```
Only available in debug mode.

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+        type: boolean
+      ip:
+        description: IP address of node
+        type: string
```
Can't this be an array of bytes?

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+    type: object
+    properties:
+      from:
+        "$ref": "#/definitions/Labels"
```
Does this still allows having a set of labels and/or an identity context? Should we extend this to allow a query to be performed against a source endpoint and a destination endpoint? or source container ID and a destination container ID?

tgraf replied:
Let's keep it as simple as possible. The smaller the API, the better.

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+    description: Layer 4 address
+    type: object
+    required:
+    - ip
```
Why isn't port required?

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+    description: Service backend address
+    type: object
+    required:
+    - ip
```
Why isn't port required?

tgraf replied:
Because we support L3 loadbalancing as well.

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+    type: object
+    properties:
+      id:
+        description: Unique identification
```
Global unique identification?

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+        description: List of endpoint IDs to consider as backends
+        type: array
+        items:
+          type: integer
```
You have previously specified the endpoint IDs as `string`

tgraf replied:
Right, I'll remove this altogether for now as it is not implemented yet.

aanm commented on api/v1/openapi.yaml:
```diff
@@ -0,0 +1,953 @@
+        items:
+          "$ref": "#/definitions/BackendAddress"
+      backend-endpoints:
+        description: List of endpoint IDs to consider as backends
```
Do we use this now?

tgraf replied:
Nope

aanm commented on cli/endpoint.go:
```diff
@@ -312,30 +235,20 @@ func configEndpoint(ctx *cli.Context) {
 	opts := ctx.Args().Tail()
 	if len(opts) == 0 {
-		ep.Opts.Dump()
+		fmt.Printf("%+v\n", cfg)
```
Can we have a "%s" plus a "String()" function?

aanm commented on cli/endpoint.go:
```diff
@@ -195,103 +166,55 @@ func verifyArguments(ctx *cli.Context) error {
-		if lbls != nil {
-			printEndpointLabels(lbls)
-		} else {
-			fmt.Printf("(Endpoint with empty labels)")
```
Bookmarked for a comment below

aanm commented on cli/endpoint.go:
```diff
@@ -515,18 +427,18 @@ func dumpEndpoints(ctx *cli.Context) {
 
-			if len(ep.SecLabel.Labels) == 0 {
-				dumpFirstEndpoint(w, &ep, seclabel, "no labels")
+			if len(ep.Identity.Labels) == 0 {
```
Why are you considering an identity with labels len == 0 but not on the comment I've bookmarked previously with "Bookmarked for a comment below"?

aanm commented on cli/endpoint.go:
```diff
@@ -540,39 +452,25 @@ func dumpEndpoints(ctx *cli.Context) {
+	if err := client.EndpointDelete(id); err != nil {
+		fmt.Fprintf(os.Stderr, "Error while detaching endpoint %s on daemon: %s\n", id, err)
+	} else {
+		fmt.Printf("Endpoint %s's successfully detached\n", id)
```
Remove the `'s` not sure why we had it in the first place. The sentence "Endpoint 20100's successfully detached" doesn't make sense to me.

aanm commented on cli/main.go:
```diff
@@ -16,23 +16,27 @@
+	log    = l.MustGetLogger("cilium-cli")
 )
 
+func Fatal(msg string, args ...interface{}) {
```
doesn't log.Fatal do this already? (I guess not and can send the msg to a different place than os.Stderr...) Either way can you replace `os.Exit(1)` with `log.Fatal(...` this way we make sure it's logged and the log.Fatal calls the os.Exit.

aanm commented on common/addressing/ip.go:
```diff
@@ -29,7 +29,10 @@ type CiliumIP interface {
 	State() uint16
 	EndpointID() uint16
 	IPNet(ones int) *net.IPNet
+	EndpointPrefix() *net.IPNet
```
Can you had some field comments here? I'm starting to get lost. (Also EndpointID() returns a uint16 and not a string, should we address this?)

aanm commented on common/addressing/ip.go:
```diff
@@ -120,6 +130,10 @@ func (ip CiliumIPv6) IPNet(ones int) *net.IPNet {
 }
 
+func (ip CiliumIPv6) EndpointPrefix() *net.IPNet {
+	return ip.IPNet(128)
```
there's a constant for this `common/addressing.go:ContainerIPv6Mask`

aanm commented on common/addressing/ip.go:
```diff
@@ -198,10 +219,19 @@ func (ip CiliumIPv4) IPNet(ones int) *net.IPNet {
 }
 
+func (ip CiliumIPv4) EndpointPrefix() *net.IPNet {
+	return ip.IPNet(32)
```
`ContainerIPv4Mask` here

aanm commented on common/addressing/ip.go:
```diff
@@ -198,10 +219,19 @@ func (ip CiliumIPv4) IPNet(ones int) *net.IPNet {
 	return net.IP(ip)
 }
 
+func (ip CiliumIPv4) State() uint16 {
```
Consistency of `State` vs `RevNATID`?

tgraf replied:
revnat is only one use of address state.

aanm commented on common/plugins/ipam.go:
```diff
@@ -0,0 +1,119 @@
+			{
+				Prefix: net.IPNet{
+					IP:   net.IPv6zero,
+					Mask: net.CIDRMask(0, 128),
```
`addressing/defaults.go:IPv6DefaultRoute`

aanm commented on common/plugins/ipam.go:
```diff
@@ -0,0 +1,119 @@
+			{
+				Prefix: net.IPNet{
+					IP:   ip,
+					Mask: net.CIDRMask(128, 128),
```
`addressing/defaults.go:ContainerIPv6Mask`?

aanm commented on common/plugins/ipam.go:
```diff
@@ -0,0 +1,119 @@
+			{
+				Prefix: net.IPNet{
+					IP:   ip,
+					Mask: net.CIDRMask(32, 32),
```
ContainerIPv4Mask

aanm commented on common/plugins/ipam.go:
```diff
@@ -0,0 +1,119 @@
+			{
+				Prefix: net.IPNet{
+					IP:   net.IPv4zero,
+					Mask: net.CIDRMask(0, 32),
```
IPv4DefaultRoute

aanm commented on common/types/loadbalancer.go:
```diff
@@ -57,17 +83,30 @@ type RevNATMap map[ServiceID]L3n4Addr
 type LoadBalancer struct {
 	BPFMapMU  sync.RWMutex
 	SVCMap    SVCMap
+	SVCMapID  SVCMapID
```
With this new MAP it should be better to have a function that atomically adds and removes a service for both SVCMap and SVCMapID, no?

tgraf replied:
Feel free to fix it up afterwards. We have to limit the scope of this PR somewhere. There are several issues that need to be addressed in this regard.

aanm commented on daemon/daemon/daemon.go:
```diff
@@ -61,27 +64,24 @@ var (
+	kvClient          kvstore.KVClient
+	containers        map[string]*types.Container
+	containersMU      sync.RWMutex
+	endpoints         map[uint16]*endpoint.Endpoint
```
1) This uint16 will become a string.

2) Maybe add a TODO for this: sense we have different prefixes for each ID (`docker:` vs `cilium:` etc) what about removing these 3 maps all together and having multiple keys pointing for the same value?

tgraf replied:
We can do this transformation in a separate PR. I'll add a TODO comment.

aanm commented on daemon/daemon/docker_watcher.go:
```diff
@@ -200,136 +201,92 @@ func (d *Daemon) updateProbeLabels(dockerID string) (bool, *types.Container, err
+		return false, nil, nil
+	}
+
+	if !d.EndpointExists(endpoint.DockerUUIDPrefix + ":" + c.ID) {
```
1) s/DockerUUIDPrefix/ContainerRuntimeUUIDPrefix?
2) A function instead would be better for consistency. `endpoint.GetCRUUID(c.ID)`?

tgraf replied:
I'm not exactly sure how to do this long term hence the open coding. I think this should be done as we converge onto a single hash map with a single lookup function. Not sure yet.

aanm commented on daemon/daemon/docker_watcher.go:
```diff
@@ -200,136 +201,92 @@ func (d *Daemon) updateProbeLabels(dockerID string) (bool, *types.Container, err
 	dockerID := container.ID
 
-	secCtxlabels, isNewLabel, err := d.PutLabels(container.OpLabels.EndpointLabels, dockerID)
+	// Remove the old labels entry
```
Even if old labels entry are the same new labels?

tgraf replied:
The logic to check for label changes is in `func (h *putEndpointIDLabels) Handle`

aanm reviewed (changes requested):
I'll continue later

Note to self: checked until daemon/daemon/endpoint.go

aanm commented on daemon/daemon/endpoint.go:
```diff
@@ -125,73 +129,82 @@ func (d *Daemon) setEndpointSecLabel(endpointID *uint16, dockerID, dockerEPID st
-func (d *Daemon) EndpointGetByDockerID(dockerID string) (*endpoint.Endpoint, error) {
-	d.endpointsMU.RLock()
-	defer d.endpointsMU.RUnlock()
+func (d *Daemon) lookupEndpointLocked(id string) (*endpoint.Endpoint, *apierror.ApiError) {
```
Yeah with all 3 maps joined this function will become simpler.

aanm commented on daemon/daemon/endpoint.go:
```diff
@@ -125,73 +129,82 @@ func (d *Daemon) setEndpointSecLabel(endpointID *uint16, dockerID, dockerEPID st
-		return ep.DeepCopy(), nil
+	eps := []*models.Endpoint{}
+	for k := range h.d.endpoints {
+		eps = append(eps, h.d.endpoints[k].GetModel())
```
```
for _, ep := range h.d.endpoints {
       eps = append(eps, ep.GetModel())
}
```

aanm commented on daemon/daemon/endpoint.go:
```diff
@@ -125,73 +129,82 @@ func (d *Daemon) setEndpointSecLabel(endpointID *uint16, dockerID, dockerEPID st
-		delete(d.endpointsDockerEP, ep.DockerEndpointID)
-		delete(d.endpoints, endpointID)
+	if ep, err := h.d.lookupEndpointLocked(params.ID); err != nil {
+		return err
```
No `NewGetEndpointError()`? Is the Error structure created by the caller?

tgraf replied:
Yes, see `lookupEndpointLocked()` implementation. The common error codes all match from all users of the function.

aanm commented on daemon/daemon/endpoint.go:
```diff
@@ -202,42 +215,88 @@ func (d *Daemon) createBPFMAPs(epID uint16) error {
+	}
+
+	h.d.endpointsMU.Lock()
+	existingEp, ok := h.d.endpoints[ep.ID]
```
This won't work the way you're expecting. Someone can access the map and change the `existingEp` since it's a pointer to the endpoint. I've seen where you use the `existingEp`. Perhaps it's better to do it like this:
```
	h.d.endpointsMU.Lock()
	existingEp, ok := h.d.endpoints[ep.ID]
        if !ok {
	        h.d.endpointsMU.Unlock()
		return NewPutEndpointIDNotFound()
	}
        epState := existingEp.State
	h.d.endpointsMU.Unlock()

	if ep.State == endpoint.StateJoining {
		if epState != endpoint.StateCreated {
			return NewPutEndpointIDInvalidState()
		}
...
```

tgraf replied:
True, this was unfinished, the new code is now locking throughout all changes. The existing code pre this PR operated on a copy which was racy as well. There was nothing to guarantee that no deletion request was processed between unlocking and locking again.

aanm commented on daemon/daemon/endpoint.go:
```diff
@@ -202,42 +215,88 @@ func (d *Daemon) createBPFMAPs(epID uint16) error {
+}
+
+func (d *Daemon) DeleteEndpointLocked(ep *endpoint.Endpoint) int {
+	errors := 0
```
Add a TODO here to return the relevant errors to the caller.
Will the user be able to clean it manually afterwards? (Talking about removing a possible dangling IP or the policy map file)

aanm commented on daemon/daemon/endpoint.go:
```diff
@@ -202,42 +215,88 @@ func (d *Daemon) createBPFMAPs(epID uint16) error {
+		h.d.conf.OptsMU.RUnlock()
+
+		ep.State = endpoint.StateReady
+		h.d.InsertEndpoint(ep)
```
Should we change also this from `InsertEndpoint` to `InsertEndpointUnlocked` and `insertEndpoint` to `insertEndpointLocked`? (Same way there's a `lookupEndpointLocked`)

aanm commented on daemon/daemon/endpoint.go:
```diff
@@ -247,180 +306,281 @@ func (d *Daemon) EndpointLeave(epID uint16) error {
 		update = true
 		if ep.SecLabel == nil {
-			ep.SecLabel.Labels = labels
+			ep.SecLabel.Labels = addLabels
```
The `SecLabel` is the `Identity` right?

tgraf replied:
Yes, we haven't changed all internal uses of this. The API uses identity now.

aanm commented on daemon/daemon/ipam.go:
```diff
@@ -16,271 +16,156 @@
-			return d.ipamConf.IPv4Allocator.Release(ip)
-		} else {
-			return nil
+	family := swag.StringValue(params.Family)
```
`family = smallCase(family)`

aanm commented on daemon/daemon/loadbalancer.go:
```diff
@@ -95,20 +94,16 @@ func (d *Daemon) svcAdd(feL3n4Addr types.L3n4AddrID, beL3n4Addr []types.L3n4Addr
-		BES: beL3n4AddrCpy,
+		FE:     feL3n4Addr,
+		BES:    beL3n4AddrCpy,
+		Sha256: feL3n4Addr.L3n4Addr.SHA256Sum(),
```
s/Sha256/SHA256

tgraf replied:
I really dislike all cap names. They are completely unreadable to me. Let's do proper camel case.

aanm commented on daemon/daemon/loadbalancer.go:
```diff
@@ -117,101 +112,133 @@ func (d *Daemon) svcAdd(feL3n4Addr types.L3n4AddrID, beL3n4Addr []types.L3n4Addr
+}
+
+func (d *Daemon) svcDelete(svc *types.LBSVC) error {
+	var svcKey lbmap.ServiceKey
```
No `d.loadBalancer.BPFMapMU.Lock()`?

aanm commented on daemon/daemon/loadbalancer.go:
```diff
@@ -117,101 +112,133 @@ func (d *Daemon) svcAdd(feL3n4Addr types.L3n4AddrID, beL3n4Addr []types.L3n4Addr
+
+func (h *getServiceID) Handle(params GetServiceIDParams) middleware.Responder {
+	d := h.daemon
+	if svc, ok := d.loadBalancer.SVCMapID[types.ServiceID(params.ID)]; ok {
```
No `d.loadBalancer.BPFMapMU.RLock()`?

tgraf replied:
Fixed

aanm commented on daemon/daemon/policy.go:
```diff
@@ -195,91 +226,165 @@ func (d *Daemon) policyAdd(path string, node *policy.Node) (bool, error) {
-			d.triggerPolicyUpdates([]policy.NumericIdentity{})
-		}
-	}()
+	defer d.policy.Mutex.Unlock()
```
Do we need to have the policy tree locked if we perform `d.triggerPolicyUpdates`?

tgraf replied:
It needs to be unlocked. Fixed everywhere. Policy update handlers will lock the tree sporadically. Handling of policy updates may become a lengthy operation and we should not delay further policy modifications. However, we should abort and restart on policy changes if we detect changes while processing old ones.

aanm commented on pkg/client/ipam.go:
```diff
@@ -0,0 +1,53 @@
+)
+
+const (
+	AddressFamilyIPv6 = "ipv6"
```
1) This is a common const between the daemon and the client right?

2) Maybe create a new type?
```
type AddressFamily string
const(
    AddressFamilyIPv6 = AddressFamily("ipv6")
...
```
that way `IPAMAllocate(family string)` becomes `IPAMAllocate(family AddressFamily)`

tgraf replied:
I'm not happy with this yet. I don't want to the client to use cilium internal types either. Can be cleaned up later, this is not bad for now.

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -153,6 +168,79 @@ type Endpoint struct {
 	PolicyMap        *policymap.PolicyMap  `json:"-"`
 	Opts             *option.BoolOptions   `json:"options"` // Endpoint bpf options.
 	Status           *EndpointStatus       `json:"status,omitempty"`
+	State            string
```
Why noy using the Status parameter?

tgraf replied:
Because Status is an array for now and state is meant to be a single state. An endpoint can only be in a single state. We can merge both together into a single type and rename status to LastStateTransitions or something but the current status is not suitable for state.

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -315,10 +403,10 @@ func OptionChanged(key string, value bool, data interface{}) {
 }
 
-func (e *Endpoint) ApplyOpts(opts option.OptionMap) bool {
+func (e *Endpoint) ApplyOpts(opts map[string]string) bool {
```
Why the `option.OptionMap` disappeared?

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -315,10 +403,10 @@ func OptionChanged(key string, value bool, data interface{}) {
 	// We want to be notified if the packets are dropped with the learn traffic
-	if val, ok := opts[OptionLearnTraffic]; ok && val {
-		opts[OptionDropNotify] = true
+	if val, ok := opts[OptionLearnTraffic]; ok && val == "enabled" {
```
Just to be safe, you're already changed all "true" to "enabled" before calling this function, right?

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -484,19 +562,45 @@ func (e *Endpoint) LogStatusOK(msg string) {
+
+func (e *Endpoint) CreateDirectory() error {
+	lxcDir := filepath.Join(".", strconv.Itoa(int(e.ID)))
+	if err := os.MkdirAll(lxcDir, 0777); err != nil {
```
`0777` seems too aggressive no?

aanm commented on plugins/cilium-docker/driver/driver.go:
```diff
@@ -213,101 +252,93 @@ func (driver *driver) createEndpoint(w http.ResponseWriter, r *http.Request) {
-	log.Debugf("Created Endpoint: %+v", endpoint)
-
-	log.Infof("New endpoint %s with IPv6: %s, IPv4: %s", endID, ipv6Address, ipv4Address)
+	log.Infof("New endpoint: docker-id=%s", endID)
```
I believe it's `docker-net-endpoint` and not `docker-id`

tgraf replied:
Fixed.

aanm commented on plugins/cilium-docker/driver/driver.go:
```diff
@@ -446,7 +446,7 @@ func (driver *driver) leaveEndpoint(w http.ResponseWriter, r *http.Request) {
 	log.Debugf("Leave request: %+v", &l)
 
-	if err := driver.client.EndpointLeaveByDockerEPID(l.EndpointID); err != nil {
+	if err := driver.client.EndpointDelete(fmt.Sprintf("docker-net-endpoint:%d", l.EndpointID)); err != nil {
```
1) Isn't endpointID a string?
2) Isn't the wrong prefix?

tgraf replied:
Fixed

aanm commented on plugins/cilium-docker/driver/ipam.go:
```diff
@@ -53,6 +52,27 @@ func (driver *driver) getDefaultAddressSpaces(w http.ResponseWriter, r *http.Req
+			PoolID: PoolIPv6,
+			Pool:   addr.IPV6.AllocRange,
+			Data: map[string]string{
+				"com.docker.network.gateway": addr.IPV6.IP + "/128",
```
We have the const with the `/128`, can we use them somehow?

aanm commented on plugins/cilium-docker/driver/ipam.go:
```diff
@@ -53,6 +52,27 @@ func (driver *driver) getDefaultAddressSpaces(w http.ResponseWriter, r *http.Req
+			PoolID: PoolIPv4,
+			Pool:   "0.0.0.0/0",
+			Data: map[string]string{
+				"com.docker.network.gateway": addr.IPV4.IP + "/32",
```
We have the const with the `/32`, can we use them somehow?

tgraf replied:
I'm considering to include the prefix in the daemon conf instead so we don't have logic in the plugins at all.

aanm commented on plugins/cilium-docker/driver/ipam.go:
```diff
@@ -96,27 +108,40 @@ func (driver *driver) requestAddress(w http.ResponseWriter, r *http.Request) {
+		if family != client.AddressFamilyIPv6 {
+			sendError(w, "Requested IPv4, received IPv6 address", http.StatusInternalServerError)
 		}
+		resp.Address = addr.IPV6 + "/128"
```
We have the const with the `/128`, can we use them somehow?

aanm commented on plugins/cilium-docker/driver/ipam.go:
```diff
@@ -96,27 +108,40 @@ func (driver *driver) requestAddress(w http.ResponseWriter, r *http.Request) {
+		if family != client.AddressFamilyIPv4 {
+			sendError(w, "Requested IPv6, received IPv4 address", http.StatusInternalServerError)
+		}
+		resp.Address = addr.IPV4 + "/32"
```
We have the const with the `/32`, can we use them somehow?

tgraf committed:
Add & update dependencies for OpenAPI code generation

This commit pulls in:
	github.com/asaskevich/govalidator
	github.com/go-openapi/analysis
	github.com/go-openapi/errors
	github.com/go-openapi/jsonpointer
	github.com/go-openapi/loads
	github.com/go-openapi/jsonreference
	github.com/go-openapi/runtime
	github.com/go-openapi/spec
	github.com/go-openapi/strfmt
	github.com/go-openapi/swag
	github.com/go-openapi/validate
	github.com/jessevdk/go-flags
	github.com/mitchellh/mapstructure
	github.com/tylerb/graceful

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
v1beta OpenAPI-2.0 specification

Adds the swagger specification for what will become the public Cilium
v1 API. It is currently namespaced in "v1beta" and will be renamed to
"v1" after initial feedback and once we are confident that no major
flaws have been introduced.

In comparison with the existing API, the following summary covers the
major changes:
 - All return codes are clearly documented.
 - The identity API has become read-only.
 - Identification of services is now done via the services ID instead
   of the hash
 - Some PUT methods which modify existing content have been converted
   to PATCH methods.
 - Some debugging API endpoints have been removed.

Signed-off-by:  <thomas@cilium.io>

tgraf committed:
Add automatically generated code based on API specification

This adds the code which is auto-generated by go-swagger. It includes:
 - Models for all data structures defined by the spec [api/v1/models/]
 - Client side code to use the API [api/v1/client]
 - Server side handling code to use the API [api/v1/server]

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
Use automatically generated API code

This commit converts Cilium to use the automatically generated code
to use the API for both the client and server side. It allows to
remove all the manually written code to implement the HTTP server,
perform API routing and manual validation of data structures provided.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
Add apierror package

Simple package to handle API errors to contain both the HTTP status
code to be returned and the error message to be exposed via Error().

Signed-off-by: Thomas Graf <thomas@cilium.io>
